### PR TITLE
Log InsertRowsNode in WAL

### PR DESCRIPTION
**Pull request by HTHou**
## Description

This PR is adding a new InsertNode type InsertRowsNode in WAL. In previous versions, the InsertRecords request will split to InsertRowNode and then log into the WAL. However, we can group some InsertRowNodes to a large InsertRowsNode. The efficiency of WAL will increase after this optimization. 

Details see https://apache-iotdb.feishu.cn/docx/FSXodTETyolOOtxdSiFcMfuznbe

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/dataregion/realtime/epoch/TsFileEpochManager.java [hunk 1]
```diff
@@ -79,7 +86,24 @@ public PipeRealtimeEvent bindPipeInsertNodeTabletInsertionEvent(
     return new PipeRealtimeEvent(
         event,
         epoch,
-        Collections.singletonMap(node.getDevicePath().getFullPath(), node.getMeasurements()),
+        node instanceof InsertRowsNode
+            ? getDevice2MeasurementsMapFromInsertRowsNode((InsertRowsNode) node)
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/dataregion/realtime/epoch/TsFileEpochManager.java:90]: Can be replaced with "return insertRowNodeList.stream().collect(Collectors.toMap(insertRowNode -> insertRowNode.getDevicePath().getFullPath(), InsertNode::getMeasurements));"

Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/dataregion/realtime/epoch/TsFileEpochManager.java:90]: Can be replaced with "insertRowNodeList.stream()
        .collect(
            Collectors.toMap(
                insertRowNode -> insertRowNode.getDevicePath().getFullPath(),
                InsertNode::getMeasurements,
                (oldMeasurements, newMeasurements) ->
                    Stream.of(Arrays.asList(oldMeasurements), Arrays.asList(newMeasurements))
                        .flatMap(Collection::stream)
                        .distinct()
                        .toArray(String[]::new)));"

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/write/InsertRowsNode.java [hunk 2]
```diff
@@ -259,12 +262,91 @@ public <R, C> R accept(PlanVisitor<R, C> visitor, C context) {
 
   @Override
   public long getMinTime() {
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/planner/plan/node/write/InsertRowsNode.java:264]: fixed

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/dataregion/realtime/epoch/TsFileEpochManager.java (hunk 1 above)
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/dataregion/realtime/epoch/TsFileEpochManager.java:88]: fixed

**Review by Caideyipi — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java [hunk 3]
```diff
@@ -885,16 +885,14 @@ public void insert(InsertRowNode insertRowNode) throws WriteProcessException {
...
-          insertRowNode, isSequence, timePartitionId, tsFileProcessorMapForFlushing);
+      TsFileProcessor tsFileProcessor =
+          insertToTsFileProcessor(insertRowNode, isSequence, timePartitionId);
 
       // check memtable size and may asyncTryToFlush the work memtable
-      for (Map.Entry<TsFileProcessor, Boolean> entry : tsFileProcessorMapForFlushing.entrySet()) {
-        if (entry.getKey().shouldFlush()) {
-          fileFlushPolicy.apply(this, entry.getKey(), entry.getValue());
+      if (tsFileProcessor != null) {
+        if (tsFileProcessor.shouldFlush()) {
```
Comment (HeimingZ) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java:894]: Maybe `if (tsFileProcessor != null && tsFileProcessor.shouldFlush())` is better.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java [hunk 4]
```diff
@@ -1105,18 +1103,15 @@ private void tryToUpdateInsertTabletLastCache(InsertTabletNode node) {
...
       throws WriteProcessException {
     if (insertRowNode.allMeasurementFailed()) {
-      return;
+      return null;
     }
     TsFileProcessor tsFileProcessor = getOrCreateTsFileProcessor(timePartitionId, sequence);
     if (tsFileProcessor == null) {
-      return;
+      return null;
     }
```
Comment (HeimingZ) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java:1115]: Maybe we can use `if (tsFileProcessor == null || insertRowNode.allMeasurementFailed())` to join two if clauses.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/recover/WALNodeRecoverTask.java [hunk 5]
```diff
@@ -155,7 +155,8 @@ private long[] recoverLastFile() {
         WALEntry walEntry = walReader.next();
         long searchIndex = DEFAULT_SEARCH_INDEX;
         if (walEntry.getType() == WALEntryType.INSERT_TABLET_NODE
-            || walEntry.getType() == WALEntryType.INSERT_ROW_NODE) {
+            || walEntry.getType() == WALEntryType.INSERT_ROW_NODE
+            || walEntry.getType() == WALEntryType.INSERT_ROWS_NODE) {
```
Comment (HeimingZ) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/recover/WALNodeRecoverTask.java:159]: There is a bug here, I forgot consider DeleteDataNode here. To fix this, please use WALEntryType.needSearch and consider the search index of DeleteDataNode.

**Review by HeimingZ — COMMENTED** [review]
It's better to add some tests about getting InsertRowsNode in the ConsensusReqReaderTest.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java (hunk 4 above)
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java]: fixed

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java (hunk 3 above)
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java]: fixed

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/recover/WALNodeRecoverTask.java (hunk 5 above)
Comment (HTHou) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/recover/WALNodeRecoverTask.java]: fixed

**Comment by HTHou** [comment]
> It's better to add some tests about getting InsertRowsNode in the ConsensusReqReaderTest.

All done. PTAL

**Review by HeimingZ — APPROVED** [review]
LGTM

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**